src/audio/mod.rs

```diff
@@ -1,39 +1,53 @@
-use crate::{ring_buffer, Pipeline, Result};
+pub mod midi;
+
+use crate::{ring_buffer, AudioUnit, Pipeline, Result};
 use anyhow::anyhow;
 use cpal::{
     traits::{DeviceTrait, HostTrait, StreamTrait},
     Device, StreamConfig,
 };
 use ringbuf::RingBuffer;
 use std::time::Duration;
+use wmidi::MidiMessage;
 
-const LATENCY: f32 = 20.0;
+const LATENCY: f32 = 50.0;
 
 pub fn run(
     input_device: &Device,
     output_device: &Device,
     config: &StreamConfig,
+    midi_port_name: Option<String>,
     mut pipeline: Pipeline,
 ) -> Result<()> {
     let latency_num_frames = (LATENCY / 1_000.0) * config.sample_rate.0 as f32;
     let latency_num_samples = latency_num_frames as usize * config.channels as usize;
 
     let ring = RingBuffer::new(latency_num_samples * 2);
     let (mut producer, mut consumer) = ring.split();
-
     ring_buffer::write_empty_samples(&mut producer, latency_num_samples)?;
 
+    let midi_messages = match midi_port_name {
+        Some(midi_port_name) => Some(midi::listen_for_input(&midi_port_name)?),
+        None => None,
+    };
+
     let input_data_fn = move |data: &[f32], _: &cpal::InputCallbackInfo| {
         if let Err(e) = ring_buffer::write_frame(&mut producer, data) {
             eprintln!("input: {:?}", e);
         }
     };
 
     let output_data_fn = move |output: &mut [f32], _: &cpal::OutputCallbackInfo| {
-        let result = ring_buffer::read_frame(&mut consumer, output.len())
-            .and_then(|frame| pipeline.process(&frame, output));
+        if let Some(midi_messages) = &midi_messages {
+            let midi_messages: Vec<MidiMessage<'static>> = midi_messages.try_iter().collect();
```

Comment:
On the right side, `midi_messages` is the receiver on the channel. `try_iter` reads every queued-up message out of the channel, and then I collect it into a `Vec`. (So this is all the messages that have arrived since the last frame was processed.)

Reply:
Does `.try_iter()` silently "consume" those messages (ie what clears those messages so that we don't re-read them again the next time)?

Reply:
Yes, it's just calling `recv()` in a loop, basically. There's no way of reading messages from a channel without consuming them.